### DISCUSSION TITLE
Berry - how to list existing rules?

### DISCUSSION THREAD

**Discussion by bogd** [arendst/Tasmota#22570] [Q&A] [answered]
I just started playing around with Berry on tasmota32, and I added some test rules. They get triggered, everything is fine, but... I cannot delete them.

Deleting them requires knowing the exact trigger used, and I cannot for the life of me remember what I typed there. And with a web browser refresh, the up-arrow history in the berry console is gone.

I have been digging through all available documentation for the last hour or so, and I cannot find a way of listing the existing rules added via `tasmota.add_rule`. How can I list the existing rules and/or delete them?

Thank you!

**Comment by sfromis** [comment, marked as answer]
You can't list them, the functions are compiled into bytecode which can't be decompiled.

You can delete using `tasmota.remove_rule`, indeed with same triggers as used for `tasmota.add_rule`, and possibly an `id` in common, if/when you have more than one add_rule for same trigger.

Of course, you can restart to get back to zero.

**Reply by sfromis** [reply]
BTW. Closest is viewing the internal list `tasmota._rules`, which is neither very readable, nor meant to be that - or even static over Tasmota releases.

**Reply by bogd** [reply]
`Of course, you can restart to get back to zero.`

This is what I was missing - I was expecting them to be persistent (similar to the "normal" rules). And this is why I was concerned about the ones I could not remove :)

Thank you!!
